src/act-compat.js

```diff
@@ -38,32 +17,101 @@ function actPolyfill(cb) {
 const act = reactAct || actPolyfill
 
 let youHaveBeenWarned = false
-// this will not avoid warnings that react-dom 16.8.0 logs for triggering
-// state updates asynchronously, but at least we can tell people they need
-// to upgrade to avoid the warnings.
-async function asyncActPolyfill(cb) {
-  // istanbul-ignore-next
-  if (
-    !youHaveBeenWarned &&
-    actSupported &&
-    reactDomSixteenPointNineIsReleased
-  ) {
-    // if act is supported and async act isn't and they're trying to use async
-    // act, then they need to upgrade from 16.8 to 16.9.
-    // This is a seemless upgrade, so we'll add a warning
-    console.error(
-      `It looks like you're using a version of react-dom that supports the "act" function, but not an awaitable version of "act" which you will need. Please upgrade to at least react-dom@16.9.0 to remove this warning.`,
-    )
-    youHaveBeenWarned = true
+let isAsyncActSupported = null
+
+function asyncAct(cb) {
+  if (actSupported === true) {
+    if (isAsyncActSupported === null) {
+      return new Promise((resolve, reject) => {
+        // patch console.error here
+        const originalConsoleError = console.error
+        console.error = function error(...args) {
+          /* if console.error fired *with that specific message* */
+          if (
+            args[0].indexOf(
+              'Warning: Do not await the result of calling ReactTestUtils.act',
+            ) === 0
+          ) {
+            // v16.8.6
+            isAsyncActSupported = false
+          } else if (
+            args[0].indexOf(
+              'Warning: The callback passed to ReactTestUtils.act(...) function must not return anything',
+            ) === 0
+          ) {
+            // no-op
+          } else {
+            originalConsoleError.call(console, args)
+          }
+        }
+        let cbReturn
+        const result = reactAct(() => {
+          cbReturn = cb()
+          return cbReturn
+        })
+
+        result.then(() => {
+          console.error = originalConsoleError
+          // if it got here, it means async act is supported
+          isAsyncActSupported = true
+          resolve()
+        }, reject)
```

Comment:
If it fails do you restore console.error?

Reply:
good catch, it doesn't! fixing.

Reply:
fixed, expanded tests to include it too